Sources/NIOHTTP1/HTTPEncoder.swift

```diff
@@ -268,6 +268,12 @@ public final class HTTPResponseEncoder: ChannelOutboundHandler, RemovableChannel
                 buffer.write(response: response)
             }, context: context, headers: response.headers, promise: promise)
         case .body(let bodyPart):
+            if case let .byteBuffer(byteBuffer) = bodyPart, byteBuffer.readableBytes == 0 {
+                // empty body parts will end the response which we don't want
+                // this happens in llhttp which is a vendored C library transpiled from TypeScript
+                // we don't want to change the upstream library and therefore handle this special case here
+                break
```

Comment:
This fix is incorrect and the analysis isn't right either.

Firstly, the analysis. Empty body parts _only_ end the response if we're using chunked encoding. That's because the empty body chunk signals the end of the body, and so is what we send in `writeTrailers`. This has nothing to do with llhttp, it's spec-behaviour.

Secondly, the fix. This leaks a promise on the outbound path. We must pass on a zero-length write down the pipeline to ensure the promise is preserved.

Reply:
AFAICT `writeTrailers` is only called on `.end` message, not on `.body` and I couldn't find any special logic that says that an empty body ends the response.

In addition this would mean our documentation is wrong an `.body` can also end a response. 
https://github.com/apple/swift-nio/blob/44a74872464ae2b8ab7722455b679d63a511b183/Sources/NIOHTTP1/HTTPTypes.swift#L128-L134
IMO this is rather unexpected and I don't think our state machines (e.g. the pipelining state machine) handle this edge case either. 

Note that I'm aware that the wire protocol ends a chunked encoding with a zero length chunk. What I'm trying to say is that llhttp is sending out the empty chunk which in turn terminates the response body. I think our swift-nio contract is different though and we only want to end a body with an explicit `.end` HTTP part, not implicitly through an empty body part. 

If we decided we want to do that we need to clearly document this and I think also audit some of our existing code. The nio code that I have written at leased assumed that `.end` must be written on a successful HTTP response.

Reply:
Oh I see, do you agree with the rest though?